irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -74,46 +75,75 @@ namespace iroha {
-        for (const auto &public_key : msg.public_keys) {
-          auto network_chain =
-              block_loader_->retrieveBlocks(height, public_key);
-
-          std::vector<std::shared_ptr<shared_model::interface::Block>> blocks;
-          network_chain.as_blocking().subscribe(
-              [&blocks](auto block) { blocks.push_back(block); });
-          if (blocks.empty()) {
-            log_->info("Downloaded an empty chain");
-            continue;
-          } else {
-            log_->info("Successfully downloaded {} blocks", blocks.size());
+        const shared_model::interface::types::HeightType start_height) {
+      const auto target_height = msg.round.block_round;
+      std::vector<std::shared_ptr<shared_model::interface::Block>> blocks;
+      auto my_height = [&] { return start_height + blocks.size(); };
+      auto expected_height = [&] { return my_height() + 1; };
+      std::unique_ptr<ametsuchi::MutableStorage> storage;
+
+      // TODO andrei 17.10.18 IR-1763 Add delay strategy for loading blocks
+      for (const auto &public_key : msg.public_keys) {
+        if (not storage) {
+          storage = getStorage().value_or(nullptr);
+          if (not storage) {
+            return boost::none;
           }
+        }
+        const auto peer_start_height = my_height();
+        auto network_chain =
+            block_loader_->retrieveBlocks(expected_height(), public_key);
+
+        auto subscription = rxcpp::composite_subscription();
+        auto subscriber = rxcpp::make_subscriber<
+            std::shared_ptr<shared_model::interface::Block>>(
+            subscription,
+            [&, this](std::shared_ptr<shared_model::interface::Block> block) {
+              log_->debug("Received block with height {}", block->height());
+              if (block->height() != expected_height()) {
+                subscription.unsubscribe();
+                log_->info("Received block with height {}, while expected {}",
+                           block->height(),
+                           expected_height());
+              } else if (not validator_->validateAndApply(block, *storage)) {
+                subscription.unsubscribe();
+                log_->info(
+                    "Could not apply block with height {} received from peer "
+                    "{}. Will retry downloading from other peers.",
+                    block->height(),
+                    public_key);
+                blocks.clear();
+                storage.reset();
+              } else {
+                log_->debug("Successfully applied block with height {}",
+                            block->height());
+                blocks.push_back(block);
+              }
+            },
+            [&, this]() {
+              log_->info("Applied {} blocks from peer {}",
+                         my_height() - peer_start_height,
+                         public_key);
+            });
+        network_chain.as_blocking().subscribe(subscriber);
```

Comment:
It might be possible to subscribe in `MutableStorage` and pass network chain directly to `ChainValidator`, and check its result.